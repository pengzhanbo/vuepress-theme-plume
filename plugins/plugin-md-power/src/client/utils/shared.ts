const cache: {
  [src: string]: Promise<void> | undefined
} = {}
export function loadScript(src: string) {
  if (__VUEPRESS_SSR__)
    return Promise.resolve()

  if (document.querySelector(`script[src="${src}"]`)) {
    if (cache[src])
      return cache[src]
    return Promise.resolve()
  }

  const script = document.createElement('script')
  script.src = src
  document.body.appendChild(script)

  cache[src] = new Promise((resolve, reject) => {
    script.onload = () => {
      resolve()
      delete cache[src]
    }
    script.onerror = reject
  })
  return cache[src]
}

export function loadStyle(href: string, target: ShadowRoot) {
  if (__VUEPRESS_SSR__)
    return

  if (target.querySelector(`link[href="${href}"]`))
    return

  const link = document.createElement('link')
  link.rel = 'stylesheet'
  link.href = href
  target.appendChild(link)
}
