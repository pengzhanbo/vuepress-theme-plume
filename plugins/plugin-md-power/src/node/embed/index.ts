import type { Markdown } from 'vuepress/markdown'
import type { MarkdownPowerPluginOptions } from '../../shared/index.js'
import { caniusePlugin, legacyCaniuse } from './caniuse.js'
import { codepenPlugin } from './code/codepen.js'
import { codeSandboxPlugin } from './code/codeSandbox.js'
import { jsfiddlePlugin } from './code/jsfiddle.js'
import { replitPlugin } from './code/replit.js'
import { pdfPlugin } from './pdf.js'
import { bilibiliPlugin } from './video/bilibili.js'
import { youtubePlugin } from './video/youtube.js'

export function embedSyntaxPlugin(md: Markdown, options: MarkdownPowerPluginOptions) {
  if (options.caniuse) {
    const caniuse = options.caniuse === true ? {} : options.caniuse
    // @[caniuse](feature_name)
    md.use(caniusePlugin, caniuse)
    // 兼容旧语法
    legacyCaniuse(md, caniuse)
  }

  if (options.pdf) {
    // @[pdf](url)
    md.use(pdfPlugin)
  }

  if (options.bilibili) {
    // @[bilibili](bvid aid cid)
    md.use(bilibiliPlugin)
  }

  if (options.youtube) {
    // @[youtube](id)
    md.use(youtubePlugin)
  }

  if (options.codepen) {
    // @[codepen](user/slash)
    md.use(codepenPlugin)
  }

  if (options.replit) {
    // @[replit](user/repl-name)
    md.use(replitPlugin)
  }

  if (options.codeSandbox) {
    // @[codesandbox](id)
    md.use(codeSandboxPlugin)
  }

  if (options.jsfiddle) {
    // @[jsfiddle](user/id)
    md.use(jsfiddlePlugin)
  }
}
