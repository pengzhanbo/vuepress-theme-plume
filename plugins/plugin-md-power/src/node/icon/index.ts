/**
 * # Icon
 *
 * ## syntax
 *
 * ::name::
 * ::provide name::
 * ::provide name =size /color extra::
 *
 * ## options
 *
 * - provide: iconify | iconfont | fontawesome
 * - prefix:
 *   - iconify: collect:name                        - prefix = "collect"
 *   - iconfont: iconfont icon-name                 - prefix = "iconfont icon-"
 *   - fontawesome: fa-solid fa-name -> fas:name    - prefix = "fas"
 *
 * - assets: url[]
 *
 * ## iconify
 *
 * - full syntax
 *   ::fluent-mdl2:toggle-filled::
 *   ::fluent-mdl2:toggle-filled =128px /#fff::
 *
 * - prefix: fluent-mdl2
 *   ::toggle-filled::
 *   ::toggle-filled =128px /#fff::
 *
 * ### iconfont
 *
 * ::name::
 * ::name =size /color::
 *
 * ### fontawesome
 *
 * ::fa-solid:name:: -> ::fas:name:: -> ::s:name:: -> ::name::
 * ::fa-brands:name:: -> ::fab:name:: -> ::b:name::
 * ::fa-regular:name:: -> ::far:name:: -> ::r:name::
 *
 * ::name fa-sm::
 *
 * @deprecated :[fluent-mdl2:toggle-filled 128px/#fff]: 此语法已废弃
 */
export * from './icon.js'
export * from './prepareIcon.js'
