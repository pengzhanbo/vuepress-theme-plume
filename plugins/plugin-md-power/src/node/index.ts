export * from '../shared/index.js'
export { resolveImageSize } from './enhance/imageSize.js'
export * from './plugin.js'
