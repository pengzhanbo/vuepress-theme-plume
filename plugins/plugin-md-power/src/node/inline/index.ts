import type { Markdown } from 'vuepress/markdown'
import type { MarkdownPowerPluginOptions } from '../../shared/index.js'
import { attrs } from '@mdit/plugin-attrs'
import { footnote } from '@mdit/plugin-footnote'
import { mark } from '@mdit/plugin-mark'
import { sub } from '@mdit/plugin-sub'
import { sup } from '@mdit/plugin-sup'
import { tasklist } from '@mdit/plugin-tasklist'
import { isPlainObject } from '@vuepress/helper'
import { abbrPlugin } from './abbr.js'
import { annotationPlugin } from './annotation.js'
import { iconPlugin } from './icons.js'
import { plotPlugin } from './plot.js'

export function inlineSyntaxPlugin(
  md: Markdown,
  options: MarkdownPowerPluginOptions,
): void {
  md.use(attrs)
  md.use(mark)
  md.use(sub)
  md.use(sup)
  md.use(footnote)
  md.use(tasklist)

  if (options.annotation) {
    /**
     * xxx [+foo] xxx
     *
     * [+foo]: xxx
     */
    md.use(annotationPlugin)
  }

  if (options.abbr) {
    /**
     * a HTML element
     *
     * [HTML]: A HTML element description
     */
    md.use(abbrPlugin)
  }

  if (options.icons) {
    // ::collect:name::
    md.use(iconPlugin, isPlainObject(options.icons) ? options.icons : {})
  }

  if (
    options.plot === true
    || (isPlainObject(options.plot) && options.plot.tag !== false)
  ) {
    // !!plot!!
    md.use(plotPlugin)
  }
}
