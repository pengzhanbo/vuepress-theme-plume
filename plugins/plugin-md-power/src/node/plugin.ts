import type { Plugin } from 'vuepress/core'
import type { MarkdownPowerPluginOptions } from '../shared/index.js'
import { addViteOptimizeDepsInclude } from '@vuepress/helper'
import { containerPlugin } from './container/index.js'
import { embedSyntaxPlugin } from './embed/index.js'
import { docsTitlePlugin } from './enhance/docsTitle.js'
import { imageSizePlugin } from './enhance/imageSize.js'
import { inlineSyntaxPlugin } from './inline/index.js'
import { prepareConfigFile } from './prepareConfigFile.js'

export function markdownPowerPlugin(
  options: MarkdownPowerPluginOptions = {},
): Plugin {
  return {
    name: 'vuepress-plugin-md-power',

    clientConfigFile: app => prepareConfigFile(app, options),

    define: {
      __MD_POWER_INJECT_OPTIONS__: options,
    },

    extendsBundlerOptions(bundlerOptions, app) {
      if (options.repl) {
        addViteOptimizeDepsInclude(
          bundlerOptions,
          app,
          ['shiki/core', 'shiki/wasm', 'shiki/engine/oniguruma'],
        )
      }
    },

    extendsMarkdown: async (md, app) => {
      docsTitlePlugin(md)
      embedSyntaxPlugin(md, options)
      inlineSyntaxPlugin(md, options)

      await containerPlugin(app, md, options)
      await imageSizePlugin(app, md, options.imageSize)
    },
  }
}
