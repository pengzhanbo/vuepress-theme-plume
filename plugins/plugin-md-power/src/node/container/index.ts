import type { App } from 'vuepress'
import type { Markdown } from 'vuepress/markdown'
import type { MarkdownPowerPluginOptions } from '../../shared/index.js'
import { isPlainObject } from '@vuepress/helper'
import { alignPlugin } from './align.js'
import { cardPlugin } from './card.js'
import { codeTabs } from './codeTabs.js'
import { demoWrapperPlugin } from './demoWrapper.js'
import { fileTreePlugin } from './fileTree.js'
import { langReplPlugin } from './langRepl.js'
import { npmToPlugins } from './npmTo.js'
import { stepsPlugin } from './steps.js'
import { tabs } from './tabs.js'

export async function containerPlugin(
  app: App,
  md: Markdown,
  options: MarkdownPowerPluginOptions,
) {
  // ::: left / right / center / justify
  alignPlugin(md)
  // ::: tabs
  tabs(md)
  // ::: code-tabs
  codeTabs(md, options.codeTabs)

  // ::: demo-wrapper
  demoWrapperPlugin(md)

  // ::: steps
  stepsPlugin(md)

  // ::: card / card-grid
  cardPlugin(md)

  if (options.npmTo) {
    // ::: npm-to
    npmToPlugins(md, typeof options.npmTo === 'boolean' ? {} : options.npmTo)
  }

  if (options.repl)
    // ::: rust-repl / go-repl / kotlin-repl
    await langReplPlugin(app, md, options.repl)

  if (options.fileTree) {
    // ::: file-tree
    fileTreePlugin(md, isPlainObject(options.fileTree) ? options.fileTree : {})
  }
}
