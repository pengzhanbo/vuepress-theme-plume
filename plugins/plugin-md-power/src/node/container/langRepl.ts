import type markdownIt from 'markdown-it'
import type Token from 'markdown-it/lib/token.mjs'
import type { App } from 'vuepress/core'
import type { ReplEditorData, ReplOptions } from '../../shared/index.js'
import container from 'markdown-it-container'
import { fs, getDirname, path } from 'vuepress/utils'

const RE_INFO = /^(#editable)?(.*)$/

function createReplContainer(md: markdownIt, lang: string) {
  const type = `${lang}-repl`
  const validate = (info: string): boolean => info.trim().startsWith(type)

  const render = (tokens: Token[], index: number): string => {
    const token = tokens[index]
    const info = token.info.trim().slice(type.length).trim() || ''
    // :::lang-repl#editable title
    const [, editable, title] = info.match(RE_INFO) ?? []

    if (token.nesting === 1)
      return `<CodeRepl ${editable ? 'editable' : ''} title="${title || `${lang} playground`}">`

    else
      return '</CodeRepl>'
  }

  md.use(container, type, { validate, render })
}

export async function langReplPlugin(app: App, md: markdownIt, {
  theme,
  go = false,
  kotlin = false,
  rust = false,
}: ReplOptions) {
  if (kotlin) {
    createReplContainer(md, 'kotlin')
  }
  if (go) {
    createReplContainer(md, 'go')
  }
  if (rust) {
    createReplContainer(md, 'rust')
  }
  theme ??= { light: 'github-light', dark: 'github-dark' }

  const data: ReplEditorData = { grammars: {} } as ReplEditorData

  const themesPath = getDirname(import.meta.resolve('tm-themes'))
  const grammarsPath = getDirname(import.meta.resolve('tm-grammars'))

  const readTheme = (theme: string) => read(path.join(themesPath, 'themes', `${theme}.json`))
  const readGrammar = (grammar: string) => read(path.join(grammarsPath, 'grammars', `${grammar}.json`))

  if (typeof theme === 'string') {
    data.theme = await readTheme(theme)
  }
  else {
    data.theme = await Promise.all([
      readTheme(theme.light),
      readTheme(theme.dark),
    ]).then(([light, dark]) => ({ light, dark }))
  }

  if (kotlin)
    data.grammars.kotlin = await readGrammar('kotlin')

  if (go)
    data.grammars.go = await readGrammar('go')

  if (rust)
    data.grammars.rust = await readGrammar('rust')

  await app.writeTemp(
    'internal/md-power/replEditorData.js',
    `export default ${JSON.stringify(data, null, 2)}`,
  )
}

async function read(file: string): Promise<any> {
  try {
    const content = await fs.readFile(file, 'utf-8')
    return JSON.parse(content)
  }
  catch {}
  return undefined
}
