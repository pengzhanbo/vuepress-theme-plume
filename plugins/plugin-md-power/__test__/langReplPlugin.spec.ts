import type { App } from 'vuepress/core'
import type { ReplOptions } from '../src/shared/repl.js'
import { dirname } from 'node:path'
import { path } from '@vuepress/utils'
import { resolveModule } from 'local-pkg'
import MarkdownIt from 'markdown-it'
import { fs, vol } from 'memfs'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import { langReplPlugin } from '../src/node/container/langRepl.js'

vi.mock('node:fs', () => fs)
vi.mock('node:fs/promises', () => fs.promises)

beforeEach(() => vol.reset())
afterEach(() => vol.reset())

const FENCE = '```'

const themesPath = dirname(resolveModule('tm-themes', {
  paths: [import.meta.url],
}) || '')
const grammarsPath = dirname(resolveModule('tm-grammars', {
  paths: [import.meta.url],
}) || '')

async function createMarkdown(options: ReplOptions = {}) {
  const md = new MarkdownIt()
  md.options.highlight = (str, lang) => `<div class="language-${lang}"><pre><code>${str}</code></pre></div>`
  const app = {
    writeTemp: vi.fn((filepath: string, content: string) => {
      filepath = path.join('/', filepath)
      fs.mkdirSync(dirname(filepath), { recursive: true })
      fs.writeFileSync(filepath, content)
    }),
  } as unknown as App

  await langReplPlugin(app, md, options)

  return { md, app }
}

function initFs() {
  vol.fromJSON({
    [path.join(themesPath, 'themes/github-light.json')]: '{ "theme": "github-light" }',
    [path.join(themesPath, 'themes/github-dark.json')]: '{ "theme": "github-dark" }',
    [path.join(grammarsPath, 'grammars/go.json')]: '{ "language": "go" }',
    [path.join(grammarsPath, 'grammars/kotlin.json')]: '{ "language": "kotlin" }',
    [path.join(grammarsPath, 'grammars/rust.json')]: '{ "language": "rust" }',
  })
}

describe('langReplPlugin', () => {
  const outputFile = '/internal/md-power/replEditorData.js'

  const code = `\
::: go-repl
${FENCE}go
const a = 1
${FENCE}
:::

::: go-repl editable
${FENCE}go
const a = 1
${FENCE}
:::

::: kotlin-repl
${FENCE}kotlin
const a = 1
${FENCE}
:::

::: kotlin-repl editable
${FENCE}kotlin
const a = 1
${FENCE}
:::

::: rust-repl
${FENCE}rust
const a = 1
${FENCE}
:::

::: rust-repl editable
${FENCE}rust
const a = 1
${FENCE}
:::

::: rust-repl title="title"
${FENCE}rust
const a = 1
${FENCE}
:::

::: rust-repl editable title="title"
${FENCE}rust
const a = 1
${FENCE}
:::

`
  it('should work with default options', async () => {
    initFs()
    const { md, app } = await createMarkdown()

    expect(md.render(code)).toMatchSnapshot()
    expect(app.writeTemp).toBeCalledWith(outputFile.slice(1), expect.any(String))
    expect(fs.readFileSync(outputFile, 'utf-8')).toMatchSnapshot()
  })

  it('should work with custom options', async () => {
    initFs()
    const { md, app } = await createMarkdown({
      go: true,
      kotlin: true,
      rust: true,
    })

    expect(md.render(code)).toMatchSnapshot()
    expect(app.writeTemp).toBeCalledWith(outputFile.slice(1), expect.any(String))
    expect(fs.readFileSync(outputFile, 'utf-8')).toMatchSnapshot()
  })

  it('should work with custom theme', async () => {
    initFs()
    const { md, app } = await createMarkdown({
      go: true,
      kotlin: true,
      rust: true,
      theme: 'vitesse-light',
    })

    expect(md.render(code)).toMatchSnapshot()
    expect(app.writeTemp).toBeCalledWith(outputFile.slice(1), expect.any(String))
    expect(fs.readFileSync(outputFile, 'utf-8')).toMatchSnapshot()
  })
})
