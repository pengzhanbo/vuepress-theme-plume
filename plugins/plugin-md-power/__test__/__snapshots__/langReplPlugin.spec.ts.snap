// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`langReplPlugin > should work with custom options 1`] = `
"<CodeRepl  title="go playground"><pre><code class="language-go"><div class="language-go"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl><CodeRepl editable title="go playground"><pre><code class="language-go"><div class="language-go"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl><CodeRepl  title="kotlin playground"><pre><code class="language-kotlin"><div class="language-kotlin"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl><CodeRepl editable title="kotlin playground"><pre><code class="language-kotlin"><div class="language-kotlin"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl><CodeRepl  title="rust playground"><pre><code class="language-rust"><div class="language-rust"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl><CodeRepl editable title="rust playground"><pre><code class="language-rust"><div class="language-rust"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl><CodeRepl  title="title"><pre><code class="language-rust"><div class="language-rust"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl><CodeRepl editable title=" title"><pre><code class="language-rust"><div class="language-rust"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl>"
`;

exports[`langReplPlugin > should work with custom options 2`] = `
"export default {
  "grammars": {
    "kotlin": {
      "language": "kotlin"
    },
    "go": {
      "language": "go"
    },
    "rust": {
      "language": "rust"
    }
  },
  "theme": {
    "light": {
      "theme": "github-light"
    },
    "dark": {
      "theme": "github-dark"
    }
  }
}"
`;

exports[`langReplPlugin > should work with custom theme 1`] = `
"<CodeRepl  title="go playground"><pre><code class="language-go"><div class="language-go"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl><CodeRepl editable title="go playground"><pre><code class="language-go"><div class="language-go"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl><CodeRepl  title="kotlin playground"><pre><code class="language-kotlin"><div class="language-kotlin"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl><CodeRepl editable title="kotlin playground"><pre><code class="language-kotlin"><div class="language-kotlin"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl><CodeRepl  title="rust playground"><pre><code class="language-rust"><div class="language-rust"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl><CodeRepl editable title="rust playground"><pre><code class="language-rust"><div class="language-rust"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl><CodeRepl  title="title"><pre><code class="language-rust"><div class="language-rust"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl><CodeRepl editable title=" title"><pre><code class="language-rust"><div class="language-rust"><pre><code>const a = 1
</code></pre></div></code></pre>
</CodeRepl>"
`;

exports[`langReplPlugin > should work with custom theme 2`] = `
"export default {
  "grammars": {
    "kotlin": {
      "language": "kotlin"
    },
    "go": {
      "language": "go"
    },
    "rust": {
      "language": "rust"
    }
  }
}"
`;

exports[`langReplPlugin > should work with default options 1`] = `
"<p>::: go-repl</p>
<pre><code class="language-go"><div class="language-go"><pre><code>const a = 1
</code></pre></div></code></pre>
<p>:::</p>
<p>::: go-repl#editable</p>
<pre><code class="language-go"><div class="language-go"><pre><code>const a = 1
</code></pre></div></code></pre>
<p>:::</p>
<p>::: kotlin-repl</p>
<pre><code class="language-kotlin"><div class="language-kotlin"><pre><code>const a = 1
</code></pre></div></code></pre>
<p>:::</p>
<p>::: kotlin-repl#editable</p>
<pre><code class="language-kotlin"><div class="language-kotlin"><pre><code>const a = 1
</code></pre></div></code></pre>
<p>:::</p>
<p>::: rust-repl</p>
<pre><code class="language-rust"><div class="language-rust"><pre><code>const a = 1
</code></pre></div></code></pre>
<p>:::</p>
<p>::: rust-repl#editable</p>
<pre><code class="language-rust"><div class="language-rust"><pre><code>const a = 1
</code></pre></div></code></pre>
<p>:::</p>
<p>::: rust-repl title</p>
<pre><code class="language-rust"><div class="language-rust"><pre><code>const a = 1
</code></pre></div></code></pre>
<p>:::</p>
<p>::: rust-repl#editable title</p>
<pre><code class="language-rust"><div class="language-rust"><pre><code>const a = 1
</code></pre></div></code></pre>
<p>:::</p>
"
`;

exports[`langReplPlugin > should work with default options 2`] = `
"export default {
  "grammars": {},
  "theme": {
    "light": {
      "theme": "github-light"
    },
    "dark": {
      "theme": "github-dark"
    }
  }
}"
`;
